posts/python_sql_engine.md

```diff
@@ -0,0 +1,187 @@
+
+Parts of SQLGlot's toolkit are being used today by [Ibis](https://github.com/ibis-project/ibis), [Quokka](https://github.com/marsupialtail/quokka), and [Splink](https://github.com/moj-analytical-services/splink).
+
+## How?
+
+There are many steps involved with actually running a simple query like:
+
+```sql
+SELECT
+  bar.a,
+  b + 'y' AS b
+FROM bar
+JOIN baz
+  ON bar.a = baz.a
+WHERE bar.a > 1
+```
+
+In this post, I'll walk through all the steps SQLGlot takes to run this query over Python objects.
+
+## Tokenizing
+
+The first step is to convert the sql string into a list of tokens. SQLGlot's tokenizer is quite simple and can be found [here](https://github.com/tobymao/sqlglot/blob/main/sqlglot/tokens.py#L763). In a while loop, it checks each character and either appends the character to the current token or makes a new token.
+
+Running the SQLGlot tokenizer shows the output.
+
+![Tokenizer Output](python_sql_engine_images/tokenizer.png)
+
+Each keyword has been converted to a SQLGlot Token object. Each token has some metadata associated with it like line / column information for error messages. Comments are also a part of the token so that comments can be preserved.
+
+## Parsing
+
+Once a SQL statement is tokenized, we don't need to worry about white space and other formatting, so it's easier to work with. We can now convert the list of tokens into an AST. The SQLGlot [parser](https://github.com/tobymao/sqlglot/blob/main/sqlglot/parser.py) is a hand written [recursive descent](https://en.wikipedia.org/wiki/Recursive_descent_parser) parser.
+
+Similar to the tokenizer, it consumes the tokens sequentially but it uses a recursive algorithm instead. The tokens are converted into a single AST node which presents the SQL query. The SQLGlot parser was designed to support various dialects so it contains many options for overriding parsing functionality.
+
+![Parser Output](python_sql_engine_images/parser.png)
+
+The AST is a generic representation of a given SQL query. Each dialect can override or implement its own generator that can convert an AST object into syntatically correct SQL.
+
+## Optimizing
+
+Once we have our AST, we need to simplify it and put it into a consistent form to actually run queries efficiently. For example, how do you know which table column 'b' comes from in `SELECT bar.a, b FROM bar JOIN baz`. Given a schema, we can infer that b comes from table baz. When optimizing queries, most engines first convert the AST into a logical plan and then optimize the plan. However, I chose to **optimize the AST directly** for two reasons.
+
+1. I wanted a way to generate 'canonical sql'. Given two SQL statements, I wanted to be able to identify if they are semantically equivalent. For example `select 2` and `select 1 + 1` are functionally the same even though the SQL query is the same. Having the optimizer output SQL would allow me to convert 1 + 1 into 2.
```

Comment:
I don't think this really captures the _why_, which might confuse people.

I think the why has to do with versioning Tobiko models. But that's hard to sum up quickly...

Perhaps we should move this after "It's easier to debug", and say something about "canonical sql" being super useful in other use cases, separate from the executor. E.g. determining if two queries are semantically equivalent.